File: Dp/Climbind Stairs/ans.py
class Solution:
    def climbStairs(self, n: int) -> int:
        return self.noOfWays(0,n, {})
    
    def noOfWays(self,currentStair,targetStair, memo):
        count = 0
        if currentStair == targetStair:
            return 1
            
        if currentStair > targetStair:
            return 0;
        currentKey = currentStair
        
        if memo.get(currentKey):
            return memo[currentKey]
        
        oneStep = self.noOfWays(currentStair + 1,targetStair, memo )
        twoStep = self.noOfWays(currentStair + 2,targetStair, memo )
        
        memo[currentStair ] = oneStep + twoStep
        
        return oneStep + twoStep
        
        

File: Dp/Climbind Stairs/ans.java
class Solution {
    public int climbStairs(int n) {
        
        return NoOfWays(0,n, new HashMap<Integer , Integer>());
        
    }
    
    public int NoOfWays(int currentStair , int targetStair, HashMap<Integer, Integer> memo){
        
        if(currentStair == targetStair)
            return 1;
        if(currentStair > targetStair)
            return 0;
        
        int currentKey = currentStair;
        
        if(memo.containsKey(currentKey))
            return memo.get(currentKey);
        
        int oneSteps = NoOfWays(currentStair + 1, targetStair,memo);
        int secondSteps = NoOfWays(currentStair + 2, targetStair,memo);
        
        memo.put(currentKey, oneSteps + secondSteps);
        
        return oneSteps + secondSteps;
    }
}
